File: tools/datarunner/datarunner/snapshot.py
from __future__ import annotations

import hashlib
import json
import time
from pathlib import Path
from typing import Any


def sha256_file(path: Path) -> str:
    h = hashlib.sha256()
    with path.open("rb") as f:
        for chunk in iter(lambda: f.read(65536), b""):
            h.update(chunk)
    return h.hexdigest()


def build_snapshot(
    *,
    prices: list[dict[str, Any]],
    id_terminal: int,
    terminal_name: str = "",
    snapshot_type: str = "commodity",
    game_version: str = "4.10.0",
    environment: str = "LIVE",
    screenshot: Path | None = None,
    captured_at: int | None = None,
    source: str = "datarunner",
) -> dict[str, Any]:
    body: dict[str, Any] = {
        "source": source,
        "game_version": game_version,
        "environment": environment,
        "id_terminal": int(id_terminal),
        "terminal_name": terminal_name or None,
        "type": snapshot_type,
        "prices": prices,
        "captured_at": captured_at if captured_at is not None else int(time.time() * 1000),
    }
    if screenshot is not None and screenshot.is_file():
        body["screenshot_sha256"] = sha256_file(screenshot)
    return body


def dumps(snapshot: dict[str, Any]) -> str:
    return json.dumps(snapshot, indent=2)

File: tools/datarunner/src/snapshot.rs
use crate::parse::PriceRow;
use serde::{Deserialize, Serialize};
use sha2::{Digest, Sha256};
use std::fs;
use std::path::Path;
use std::time::{SystemTime, UNIX_EPOCH};

#[derive(Clone, Debug, Serialize, Deserialize)]
pub struct Snapshot {
    pub source: String,
    pub game_version: String,
    pub environment: String,
    pub id_terminal: i64,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub terminal_name: Option<String>,
    #[serde(rename = "type")]
    pub kind: String,
    pub prices: Vec<PriceRow>,
    pub captured_at: i64,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub screenshot_sha256: Option<String>,
}

pub fn sha256_file(path: &Path) -> anyhow::Result<String> {
    let bytes = fs::read(path)?;
    let mut h = Sha256::new();
    h.update(&bytes);
    Ok(hex::encode(h.finalize()))
}

pub fn now_ms() -> i64 {
    SystemTime::now()
        .duration_since(UNIX_EPOCH)
        .map(|d| d.as_millis() as i64)
        .unwrap_or(0)
}

pub fn build_snapshot(
    prices: Vec<PriceRow>,
    id_terminal: i64,
    terminal_name: &str,
    snapshot_type: &str,
    game_version: &str,
    environment: &str,
    screenshot: Option<&Path>,
) -> anyhow::Result<Snapshot> {
    let sha = match screenshot {
        Some(p) if p.is_file() => Some(sha256_file(p)?),
        _ => None,
    };
    Ok(Snapshot {
        source: "datarunner".into(),
        game_version: game_version.into(),
        environment: environment.into(),
        id_terminal,
        terminal_name: if terminal_name.is_empty() {
            None
        } else {
            Some(terminal_name.into())
        },
        kind: snapshot_type.into(),
        prices,
        captured_at: now_ms(),
        screenshot_sha256: sha,
    })
}
